### PR TITLE
issue #1: more restrictive operations

### PR DESCRIPTION
**Pull request by xgbuils**
PR for #1

### CODE REVIEW

**Comment by fibo** [comment]
Thanks for the PR,

I would merge it, however I think that code could be simplified. If you don't mind I will edit the code later.
For example making errors plain strings and exporting them.

I am also not sure what to do if a user defines something that is not a group, like the range `[0.5, 2]` with the multiplicative operation.
Checking if a group is well defined can be very tricky, I am not sure if a tiny lib (or even a computer) can do it.
Hence, I am not sure about the error messages. If I invert an element and it is not contained in the group, the message should be something like "the group is not well defined". Such a check could be avoided if we let the responsibility to give a well defined group to the user.

The `(0, 1]` multiplicative group is a good idea.

**Comment by xgbuils** [comment]
Hi @fibo,

> I would merge it, however I think that code could be simplified. If you don't mind I will edit the code later.
For example making errors plain strings and exporting them.

The problem of plain strings is that are not configurables. If you want to report diferent messages depending on the name of callback that produces the exception, then you needs one message for each operation and type of error (and maybe the wrong value passed as argument). That increases like a multiplicative factor. Of course, if you wants report messages less specific, this problem does not exist.

Another way to solve the multiplicative factor on error messages without losing specificity is using plain strings with certain format that then it can be parsed.

> I am also not sure what to do if a user defines something that is not a group, like the range [0.5, 2] with the multiplicative operation.
> Checking if a group is well defined can be very tricky, I am not sure if a tiny lib (or even a computer) can do it.
> Hence, I am not sure about the error messages. If I invert an element and it is not contained in the group, the message should be something like "the group is not well defined". Such a check could be avoided if we let the responsibility to give a well defined group to the user.

In conclusion, I think that the 3 types of managing errors that I mentioned in issue #1. The first point is the one that you wanted. It would have been nice to know before.

Of course, check if a group is well defined is really tricky. Hence, I proposed check in runtime. Another point of view is to delegate the responsibility to give well defined to the user.

**Comment by fibo** [comment]
Thank you for your contribution!

Can you also tell me which kind of experiments are you doing with algebra? We can interact via chat (skype or hangout) if you want.

**Comment by xgbuils** [comment]
You are welcome!

Currently I'm not doing any experiment with algebra. I love maths but I ended up studying computer science. Therefore currently I'm interested in this library as a concept and not as something to use. What about you? Why dedided to create `algebra` and its sublibraries? If you want, you can write to me via e-mail which I have in my github profile.

**Comment by fibo** [comment]
Sure, I wrote you an email.
The main motivation was answering the question:

> does computers understand algebra like humans?

The answer  I found is: no.

When a human studies group theory it starts from its [laws](https://en.wikipedia.org/wiki/Group_(mathematics)#Definition). I could not implement (read: *make a computer undestand algebra*) an group lib just using those laws, cause they are abstracts.

**Comment by fibo** [comment]
By now it is ok a runtime check. Another way to do it, could be expose a test function you feed with the Group definition and a test set with results of Operations, so the user can check the definition looks good.
